taui/src/selectors/network-listing-route.js

```diff
@@ -0,0 +1,80 @@
+// @flow
+import lonlat from '@conveyal/lonlat'
+import get from 'lodash/get'
+import memoize from 'lodash/memoize'
+import {createSelector} from 'reselect'
+
+import createTransitiveRoutes from '../utils/create-transitive-routes'
+import createDirectRoutes from '../utils/create-direct-routes'
+
+import selectActiveNetworkIndex from './active-network-index'
+
+/**
+ * This assumes loaded query, paths, and targets.
+ */
+const memoizedTransitiveRoutes = memoize(
+  (n, i, s, e, car) => (car ? createDirectRoutes(n, s, e) : createTransitiveRoutes(n, s, e)),
+  (n, i, s, e, car) =>
+    `${n.name}-${i}-${n.originPoint.x}-${n.originPoint.y}-${lonlat.toString(e.position)}-car-${car}`
+)
+
+const routeToString = s =>
+  s.map(s => `${s.name}-${s.backgroundColor}-${s.type}`).join('-')
+
+const uniqueSegments = routeSegments => {
+  const foundKeys = {}
+  return (routeSegments || []).reduce((uniqueRoutes, route) => {
+    const key = routeToString(route)
+    if (!foundKeys[key]) {
+      foundKeys[key] = true
+      return [...uniqueRoutes, route]
+    }
+    return uniqueRoutes
+  }, [])
+}
```

Comment:
It's kind of a bummer to have two copies of this somewhat-complex logic.  Is there somewhere we could at least put this top part, which is fully identical between them, so it could be defined once and shared?

The `createSelector` functions are a bit trickier, since there are differences in both the input object and the return structure.  Ideally we could have a shared `composeRoute` function to do the "check for completeness and build a route object" part, though it's possibly managing the differences in the inputs would be too cumbersome for that to come out well.